src/middlewared/middlewared/plugins/init_shutdown_script.py

```diff
@@ -104,33 +103,50 @@ class Config:
         if verrors:
             raise verrors
 
+    @private
+    async def execute_task(self, task):
+        task_type = task['type']
+        cmd = None
+
+        if task_type == 'COMMAND':
+            cmd = task['command']
+        elif os.path.exists(task['script'] or '') and os.access(task['script'], os.X_OK):
+            cmd = f'exec {task["script"]}'
+
+        if cmd:
+            proc = await Popen(
+                cmd,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                shell=True,
+                close_fds=True
+            )
+            await proc.communicate()
+
+            if proc.returncode:
+                self.middleware.logger.debug(
+                    f'Execution failed for {task_type} {task["command"] if task_type == "COMMAND" else task["script"]}'
+                )
+
     @private
     @accepts(
         Str('when')
     )
-    def execute_init_tasks(self, when):
-        tasks = self.middleware.call_sync(
+    @job()
+    async def execute_init_tasks(self, job, when):
+
+        tasks = await self.middleware.call(
             'initshutdownscript.query', [
                 ['enabled', '=', True],
                 ['when', '=', when]
             ])
 
         for task in tasks:
-            task_type = task['type']
-            ret = None
-
-            if task_type == 'COMMAND':
-                ret = subprocess.run(
-                    task['command'],
-                    shell=True
-                )
-            elif os.path.exists(task['script']) and os.access(task['script'], os.X_OK):
-                ret = subprocess.run(
-                    f'exec {task["script"]}',
-                    shell=True
-                )
-
-            if ret and ret.returncode:
+            try:
+                await asyncio.wait_for(self.execute_task(task), timeout=task['timeout'])
+            except asyncio.TimeoutError:
                 self.middleware.logger.debug(
-                    f'Execution failed for {task_type} {task["command"] if task_type == "COMMAND" else task["script"]}'
+                    f'{task["type"]} {task["command"] if task["type"] == "COMMAND" else task["script"]} timed out'
                 )
+
+        job.set_progress(100, f'Completed tasks for {when}')
```

Comment:
If we intend to update job progress after each script, I don't see where we do this. If we don't, why do we set it so 100% then?